TBF Header Permissions (#2172)
tock/tock: doc/TockBinaryFormat.md

    the linker. If a fixed address is not required this should be set to
    `0xFFFFFFFF`.

#### `6` Permissions

The `Permissions` section allows an app to specify driver permissions that it
is allowed to use. All driver syscalls that an app will use must be listed. The
list should not include drivers that are not being used by the app.

The data is stored in the optional `TbfHeaderV2Permissions` field. This

> includes an array of all the `perms`.

Reviewer: What is the behavior when things are not specified? The obvious backwards-compatible choice is to default to permitted, but as a security-oriented operating system, is that actually the default we want to put into the world?

---

A perhaps more nuanced version might be that if an application specifies _no_ permissions, then it has all permissions. However, once an application chooses to specify _any_ permissions, we expect it to specify all of its requested permissions?

Author: I think by default we would allow all with no permissions.

Eventually I think we could then migrate to a disabled by default once all the usersapce apps/tools have migrated.

Reviewer 2: Does "disabled by default" mean disabled if the permissions header is not present, or disabled if the permissions header is present but empty?

I would oppose a disable by default if the permissions header is not present for the following reasons:

A. This is not compatible with drivers that use all 32 bits of the command ID (i.e. have command numbers >= 64)
B. Particular projects using Tock that have their own security concerns may need their own permissions header, and that would require them to carry both sets of headers. That could waste a lot of space.

Author: Eventually I would say "disabled by default" means both, so access is disabled if there is no permissions header or if it is empty. Like I said though, I think initially we would allow everything and only once everything is working and we have had a release or two for everything to catch up would we make that switch.

In saying that, this PR has no impact on that decision, we could just always default to enable by default and have a kernel config that disables access.

Yep, this PR is limited to less then 64 command numbers. Some drivers will need to have this numbers changed (ninedef) to meet this. As discussed on the call we allow doubling the driver numbers to meet more bit masks.

The idea is this would be the header all apps use. If others want to duplicate this we can't really do anything to help that.

Reviewer 2: > Eventually I would say "disabled by default" means both, so access is disabled if there is no permissions header or if it is empty. Like I said though, I think initially we would allow everything and only once everything is working and we have had a release or two for everything to catch up would we make that switch.
> 
> In saying that, this PR has no impact on that decision, we could just always default to enable by default and have a kernel config that disables access.

If you think that discussion is outside the scope of this PR, then I am fine having it elsewhere.

I do want to point out that OpenTitan has not decided what process permissions model it needs, and we cannot determine now whether OpenTitan will use this permissions model. By default, I'd assume that OpenTitan will have its own permissions system, so we want to support that case.

```
0             2             4
+-------------+-------------+---------...--+
| Type (6)    | Length      | perms        |
+-------------+-------------+---------...--+
```

The `perms` array is made up of a number of elements of
`TbfHeaderDriverPermission`. The length of the TLV can be used to determine
the number of array elements. The elements in `TbfHeaderDriverPermission` are
described below:

`driver_number` is the number of the driver that is allowed. This for example
could be `0x00000` to indicate that the `Alarm` syscalls are allowed.

`allowed_commands` is a bit mask of the allowed commands. For example a value
of `0b0001` indicates that only command 0 is allowed. `0b0111` would indicate

> that commands 2, 1 and 0 are all allowed. Subscribe and allow commands are

Reviewer: Should `subscribe` be mass-permitted? One might imagine an information leakage channel coming from events being reported. I believe that in many cases right now, if Application A tells something to start sampling, an Application B can subscribe to the event stream without explicitly starting the sampling itself.

Author: I thought application B shouldn't get an event callback unless it started the command.

Author: Also, I'll clarify this that `allow` and `subscribe` are only permitted if the driver is permitted. If the driver is permitted all `allow` and `subscribe` calls are allowed.

Reviewer: I think that clarification is good. However, we do have implementations that don't enforce that right now, e.g. for the ADC:

```rust
            // perform callback
            self.appid.map(|id| {
                self.apps
                    .enter(*id, |app, _| {
                        app.callback.map(|callback| {
                            calledback = true;
                            callback.schedule(
                                AdcMode::ContinuousSample as usize,
                                self.channel.get(),
                                sample as usize,
                            );
                        });
                    })
```

On the upcall path, callbacks will be issued to any apps that have subscribed, regardless of which app started the ADC. That may be a bug / unintended consequence in the ADC interface, but it seems sensible that if we're going to filter defensively, we should filter against apps "listening-in" when they didn't declare their intent/ability to do so.

list should not include drivers that are not being used by the app.

The data is stored in the optional `TbfHeaderV2Permissions` field. This
includes an array of all the `perms`.

```
0             2             4
+-------------+-------------+---------...--+
| Type (6)    | Length      | perms        |
+-------------+-------------+---------...--+
```

The `perms` array is made up of a number of elements of
`TbfHeaderDriverPermission`. The length of the TLV can be used to determine
the number of array elements. The elements in `TbfHeaderDriverPermission` are
described below:

`driver_number` is the number of the driver that is allowed. This for example
could be `0x00000` to indicate that the `Alarm` syscalls are allowed.

`allowed_commands` is a bit mask of the allowed commands. For example a value
of `0b0001` indicates that only command 0 is allowed. `0b0111` would indicate
that commands 2, 1 and 0 are all allowed. Subscribe and allow commands are
always allowed as long as the specific `driver_number` has been specified. If
a `driver_number` has not been specified for the capsule driver then `allow`
and `subscribe` will be blocked.

> `driver_number` can be repeated. Each time `driver_number` is repeated the

Reviewer 3: I worry about stateful specifications like this. It places requirements on a parser to maintain that state. 

Instead, and I think this came up in the call, there should be an offset field that specifies the offset of the bits. So you can split this into 2 32-bit values, and offset and a bitmask. The value of a bit in the bitmask is `bit_position` + `offset`. I.e., if bit 7 is set and the offset value is 300, it means command 307 is allowed.

This means that dense command spaces are more efficient to permission, but you can permission arbitrarily sparse ones not too badly (unlike say with this next-64 approach, which would not like the 9dof identifiers).

It would be good to also explicitly specify the layout of the the `TbfHeaderdriverPermission` structure as well.

Author: I was hesitant to do it that way as it adds an extra 4 bytes to every driver specified, when I think it will be rare that there will be more then 64 commands (assuming we change the current drivers to all use dense numbering).

I have converted it to using the `offset` field though.

    the linker. If a fixed address is not required this should be set to
    `0xFFFFFFFF`.

#### `6` Permissions

The `Permissions` section allows an app to specify driver permissions that it
is allowed to use. All driver syscalls that an app will use must be listed. The
list should not include drivers that are not being used by the app.

The data is stored in the optional `TbfHeaderV2Permissions` field. This
includes an array of all the `perms`.

```
0             2             4
+-------------+-------------+---------...--+
| Type (6)    | Length      | perms        |
+-------------+-------------+---------...--+
```

The `perms` array is made up of a number of elements of
`TbfHeaderDriverPermission`. The length of the TLV can be used to determine
the number of array elements. The elements in `TbfHeaderDriverPermission` are
described below:

> 

Reviewer 4: ```suggestion

```text
Driver Permission Structure:
0             2             4             6             8
+-------------+-------------+---------------------------+
| driver_number             | offset                    |
+-------------+-------------+-------------+-------------+
| allowed_commands                                      |
+-------------------------------------------------------+
```

```

> `driver_number` is the number of the driver that is allowed. This for example

Reviewer 4: ```suggestion
* `driver_number` is the number of the driver that is allowed. This for example
```

could be `0x00000` to indicate that the `Alarm` syscalls are allowed.

> 

Reviewer 4: ```suggestion

* The `offset` field in `TbfHeaderDriverPermission` indicates the offset of the
`allowed_commands` bitmask. All of the examples described in the paragraph
above assume an `offset` of 0. The `offset` field indicates the start of the
`allowed_commands` bitmask. The `offset` is multiple by 64 (the size of the
`allowed_commands` bitmask). For example an `offset` of 1 and a
`allowed_commands` value of `0b0001` indicates that command 64 is allowed.

```

> `allowed_commands` is a bit mask of the allowed commands. For example a value

Reviewer 4: ```suggestion
* `allowed_commands` is a bit mask of the allowed commands. For example a value
```

of `0b0001` indicates that only command 0 is allowed. `0b0111` would indicate
that commands 2, 1 and 0 are all allowed. Note that this assumes `offset` is
0, for more details on `offset` see below.

Subscribe and allow commands are always allowed as long as the specific
`driver_number` has been specified. If a `driver_number` has not been specified
for the capsule driver then `allow` and `subscribe` will be blocked.

> The `offset` field in `TbfHeaderDriverPermission` indicates the offset of the
> `allowed_commands` bitmask. All of the examples described in the paragraph
> above assume an `offset` of 0. The `offset` field indicates the start of the
> `allowed_commands` bitmask. The `offset` is multiple by 64 (the size of the
> `allowed_commands` bitmask). For example an `offset` of 1 and a
> `allowed_commands` value of `0b0001` indicates that command 64 is allowed.

Reviewer 4: ```suggestion
```

> Multiple `offset` and `allowed_commands` can be included, so long as no

Reviewer 4: ```suggestion
Multiple `TbfHeaderDriverPermission` with the same `driver_numer` can be included, so long as no
```
